### DISCUSSION TITLE
How to perform a temporal undo?

### DISCUSSION THREAD

**Discussion by Agobin** [GrapesJS/grapesjs#3678] [Q&A]
I have a section in my application where I list all the changes made by a user and allow the user to quickly switch between states. I rely on the UndoManager's stack to switch between states. However, when I sometimes perform an undo to switch to a particular state, the stack changes.

Is there a way to temporarily perform an undo without actually changing the stack?

**Comment by artf** [comment]
Hey @Agobin I've actually started talking about the refactoring of UndoManager [here](https://github.com/artf/grapesjs/issues/3639), which should allow better support for what you're doing. 
Would you mind describing how did you actually create your list of changes? 

> undo without actually changing the stack?

I'm wondering how do you perform the undo, as calling `undoManager.undo` doesn't change the stack.

**Reply by Agobin** [reply]
Hi @artf ,

In my application, I track every action I do. And whenever I perform an action that can be undone, I add a description of the action in a list. This way, the list has all the changes made by the app but this is highly manual.

> I'm wondering how do you perform the undo, as calling undoManager.undo doesn't change the stack.

Performing an undo doesn't change the stack, however, say I have ten models in the stack relating to 10 different actions that can be undone. If I perform 5 undos, the stack doesn't change. If at this point I perform a new undoable action, it changes the stack. The stack will now have 6 models instead of 11. The 5 actions that were undone will be removed from the stack.

Currently, I used an older version of grapes and the undo feature doesn't work effectively. Sometimes when I perform an undo, it performs and undo action but the view doesn't change. I was wondering if there was a way to replace the undo manager of the older version with that of the recent version? Updating the whole library would break certain features of the application.

**Reply by artf** [reply]
> If I perform 5 undos, the stack doesn't change. If at this point I perform a new undoable action, it changes the stack. The stack will now have 6 models instead of 11. The 5 actions that were undone will be removed from the stack.

Well, this is how any UndoManager is supposed to work as you're changing something in the middle.

> Currently, I used an older version of grapes and the undo feature doesn't work effectively. Sometimes when I perform an undo, it performs and undo action but the view doesn't change. I was wondering if there was a way to replace the undo manager of the older version with that of the recent version? Updating the whole library would break certain features of the application.

I guess you should try it on your own fork. 
Did you try at least to upgrade? What version do you use? What errors are you facing?

**Reply by Agobin** [reply]
I finally upgraded my version of grapes.

**Reply by artf** [reply]
Great, anyway I'd suggest following issue #3639 
If you have any suggestions on how to make the UndoManager more extendable (eg. events, API, etc.) post it there.

**Reply by Agobin** [reply]
Okay, I will.
